src/3d/framegraph/qgsabstractrenderview.h

```diff
@@ -0,0 +1,110 @@
+/***************************************************************************
+  qgsabstractrenderview.h
+  --------------------------------------
+  Date                 : June 2024
+  Copyright            : (C) 2024 by Benoit De Mezzo
+  Email                : benoit dot de dot mezzo at oslandia dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSABSTRACTRENDERVIEW_H
+#define QGSABSTRACTRENDERVIEW_H
+
+#include "qgis_3d.h"
+
+#include <QObject>
+#include <Qt3DRender/QRenderTargetOutput>
+
+#define SIP_NO_FILE
+
+class QColor;
+class QRect;
+class QSurface;
+
+namespace Qt3DCore
+{
+  class QEntity;
+}
+
+namespace Qt3DRender
+{
+  class QRenderSettings;
+  class QCamera;
+  class QFrameGraphNode;
+  class QLayer;
+  class QViewport;
+  class QTexture2D;
+  class QSubtreeEnabler;
+  class QRenderTargetSelector;
+}
+
+class QgsFrameGraph;
+
+/**
+ * \ingroup 3d
+ * \brief Base class for 3D render view
+ *
+ * An instance of QgsAbstractRenderView is a branch (ie. a render pass, ie. a render view) of in the framegraph
+ *
+ * Will be used with QgsFrameGraph::registerRenderView.
+ *
+ * \note Not available in Python bindings
+ * \since QGIS 3.40
+ */
+class _3D_EXPORT QgsAbstractRenderView : public QObject
+{
+    Q_OBJECT
+  public:
+
+    /**
+     * Constructor for QgsAbstractRenderView with the specified \a parent object.
+     */
+    QgsAbstractRenderView( QObject *parent, const QString &viewName );
+
+    //! Set output to screen (ie. nullptr) or to a render target output
+    virtual void setTargetOutputs( const QList<Qt3DRender::QRenderTargetOutput *> &targetOutputList );
+
+    //! Returns the list of all target outputs
+    virtual QList<Qt3DRender::QRenderTargetOutput *> targetOutputs() const;
+
+    //! Updates texture sizes for all target outputs
+    virtual void updateTargetOutputSize( int width, int height );
+
+    //! Returns the 2D texture attached at the \a attachment point, if any
+    virtual Qt3DRender::QTexture2D *outputTexture( Qt3DRender::QRenderTargetOutput::AttachmentPoint attachment ) const;
+
+    //! Returns the layer to be used by entities to be included in this render view
+    Qt3DRender::QLayer *layerToFilter() const;
+
+    //! Returns the viewport associated to this render view
+    virtual Qt3DRender::QViewport *viewport() const;
```

Comment:
Thanks @wonder-sk for the review!

Sorry for the confusion, I thought I had answered your questions previsiously. As I said this PR is the first of a series, the next PR will provide the other render view extractions. If you like, you can check how the final framegraph and renderviews will look like on this [branch](https://github.com/benoitdm-oslandia/QGIS/tree/feature/extendable-framegraph). I hope the use of these functions will be clearer.

And I am agree the viewport() function is to specific but I does no agree with the others. By reading your comments I understand the functions names or/and comments could be improved. I will correct this asap.

Reply:
I am looking at the linked branch, but I still fail to understand why we need `Qt3DRender::QRenderTargetSelector *mRenderTargetSelector` member in `QgsAbstractRenderView` and all the functions for rendering targets in the base class. Why do they need to be there? Why can't render passes define their target selectors as needed? As noted above, I can imagine that we would have e.g. `handleWindowResize(QSize size)` so that render views can update their texture sizes if they depend on the window size, but I am unsure about all the other ones.

Same with layerToFilter() - each render view should provide its layer(s) for filtering if needed in their derived class, and so a render view could e.g. expose multiple layers for different occasions, or maybe no layers at all.

Reply:
> I am looking at the linked branch, but I still fail to understand why we need `Qt3DRender::QRenderTargetSelector *mRenderTargetSelector` member in `QgsAbstractRenderView` and all the functions for rendering targets in the base class. Why do they need to be there? Why can't render passes define their target selectors as needed? As noted above, I can imagine that we would have e.g. `handleWindowResize(QSize size)` so that render views can update their texture sizes if they depend on the window size, but I am unsure about all the other ones.

The goal here is to go further than simply handling resizing. This idea is to separate the responsibility between the texture creation and their usage to: ease their accesses, sets, updates size. Also, by having the `mRenderTargetSelector` in the abstract class we match the existing renderviewes : they all have one `mRenderTargetSelector`.

> 
> Same with layerToFilter() - each render view should provide its layer(s) for filtering if needed in their derived class, and so a render view could e.g. expose multiple layers for different occasions, or maybe no layers at all.

ok it is possible to do without them, I will remove the `layerToFilter()` and `viewport` function.

Reply:
> The goal here is to go further than simply handling resizing. This idea is to separate the responsibility between the texture creation and their usage to: ease their accesses, sets, updates size.

I am not sure I understand. Could you please expand your thoughts, and how that relates to having `Qt3DRender::QRenderTargetSelector *mRenderTargetSelector` in `QgsAbstractRenderView` ?

> Also, by having the mRenderTargetSelector in the abstract class we match the existing renderviewes : they all have one mRenderTargetSelector.

Well, not necessarily... How about ambient occlusion render view (in your branch) - it has _two_ render target selectors - which one is _the_ one?

My point is that we should expect arbitrary sub-graph in a single render view, and not have some frame graph nodes treated specially in the base class. If there would be always exactly one render target selector in every render view, then it would be fine, but I don't see that here.

> I will remove the layerToFilter() and viewport function.

Thanks!

Reply:
> I am not sure I understand. Could you please expand your thoughts, and how that relates to having Qt3DRender::QRenderTargetSelector *mRenderTargetSelector in QgsAbstractRenderView ?

because you need a `QRenderTargetSelector` to be able to add `QRenderTarget` to handle textures. Indeed if a renderview does not need to write into textures, it will not need all these helping functions. But in that case the instantiated object will only contain one empty pointer to the `mRenderTargetSelector`.

> Well, not necessarily... How about ambient occlusion render view (in your branch) - it has two render target selectors - which one is the one?

The one in which you want to write the final result. The other one is just for a temporary one. 

I got your point but at the beginning I had made nearly the same design you want to, but I had a lot of duplicate code between the different renderviews. In order to find a compromise, I can provide 2 abstract classes, one really basic and another one extending the first with the output texture helper functions.

Reply:
I have checked the code on your branch again (https://github.com/qgis/QGIS/compare/master...benoitdm-oslandia:QGIS:feature/extendable-framegraph) and I still struggle to understand why they should be there - they make the abstract class more complicated, they are not generic enough, and the code is less readable. I don't see where you would get a lot of duplicate code either...

Just to make sure we're on the same page, here's what I am suggesting:
- remove setTargetOutputs(), targetOutputs(), outputTexture(), onTargetOutputUpdate() functions and mRenderTargetSelector, mTargetOutputs member variables from QgsAbstractRenderView class
- rename updateTargetOutputSize(QSize) to handleWindowResize(QSize) and leave its default implementation empty
- using the forward render view as an example:
  - QgsForwardRenderView constructor
    - creation of textures and render outputs should be done here (not in QgsFrameGraph::constructForwardRenderPass())
    - it will create QRenderTarget, assign render outputs, create QRenderTargetSelector and assign it
  - QgsForwardRenderView should get two new public non-virtual functions:
    - `QTexture2D* colorTexture()`
    - `QTexture2D* depthTexture()`
   - QgsForwardRenderView would implement handleWindowResize() and set texture size of its color+depth textures
   - QgsFrameGraph::constructForwardRenderPass()
      - the whole function body would end up being just:
      `registerRenderView( new QgsForwardRenderView( this, mMainCamera ), FORWARD_RENDERVIEW );`

This is IMHO much, much easier to read and understand - because the whole frame graph fragment is built in one place together with any other objects - rather than having a bunch of virtual methods that only make sense in some contexts.